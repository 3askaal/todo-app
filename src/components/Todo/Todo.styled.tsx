import styled, { CSSObject } from 'styled-components'
import { Spacings } from '../../style'

export const STodo = styled.div(() => ({
  display: 'flex',
  alignItems: 'stretch',
  width: '100%',
}))

export const STodoCheckbox = styled.div({
  display: 'flex',
  marginRight: Spacings.s,
})

export const STodoInput = styled.div(
  ({ completed }: { completed?: boolean }): CSSObject => ({
    flexGrow: 1,
    position: 'relative',
    textDecoration: completed ? 'line-through' : undefined,

    '> * > *': {
      borderTopLeftRadius: 0,
      borderBottomLeftRadius: 0,
    },
  }),
)

export const STodoButton = styled.div({
  display: 'flex',
  marginLeft: Spacings.s,
})
