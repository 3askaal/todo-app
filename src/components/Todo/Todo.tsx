import React, { FC, ReactElement } from 'react'
import { X as XIcon } from 'react-feather'
import { STodo, STodoCheckbox, STodoInput, STodoButton } from './Todo.styled'
import { Checkbox, Input, Button } from '..'

export interface TodoProps {
  content: string
  completed: boolean
  onUpdate: (update: object) => void
  onDelete: () => void
}

export const Todo: FC<TodoProps> = ({
  content,
  completed,
  onUpdate,
  onDelete,
}: TodoProps): ReactElement => {
  return (
    <STodo data-testid="todo">
      <STodoCheckbox data-testid="todo-completed">
        <Checkbox
          checked={completed}
          onChange={event => onUpdate({ completed: event.target.checked })}
          data-testid="todo-completed-input"
        />
      </STodoCheckbox>
      <STodoInput data-testid="todo-content">
        <Input
          value={content}
          onChange={event => onUpdate({ content: event.target.value })}
          data-testid="todo-content-input"
        />
      </STodoInput>
      <STodoButton>
        <Button onClick={onDelete} data-testid="todo-button">
          <XIcon />
        </Button>
      </STodoButton>
    </STodo>
  )
}
