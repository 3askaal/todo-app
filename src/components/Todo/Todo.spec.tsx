import React from 'react'
import { Todo } from './Todo'
import {
  render,
  act,
  fireEvent,
  customBeforeEach,
  customAfterEach,
  waitForData,
} from '../../test/utils'

let container: any = null

beforeEach(() => {
  container = customBeforeEach()
})

afterEach(() => {
  customAfterEach(container)
})

describe('Todo', () => {
  it('runs', async () => {
    const tree = render(<Todo content="find new job" completed={false} />)

    await act(async () => {
      await waitForData()
    })

    expect(tree).toMatchSnapshot()
  })

  it('updates todo', async () => {
    const onUpdate = jest.fn()

    const { getByTestId } = render(
      <Todo content="find new job" completed={false} onUpdate={onUpdate} />,
    )

    await act(async () => {
      await waitForData()
    })

    const todoContentInput: any = getByTestId('todo-content-input')
    fireEvent.change(todoContentInput, {
      target: { value: 'find awesome new job' },
    })

    const todoCompletedInput: any = getByTestId('todo-completed-input')
    fireEvent.click(todoCompletedInput)

    await act(async () => {
      await waitForData(500)
    })

    expect(onUpdate).toHaveBeenCalledTimes(2)
  })
})
