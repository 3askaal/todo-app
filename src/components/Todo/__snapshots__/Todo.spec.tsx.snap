// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo runs 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div />
    <div>
      <div
        class="sc-AxjAm fwkaia"
        data-testid="todo"
      >
        <div
          class="sc-AxirZ gsjyTv"
          data-testid="todo-completed"
        >
          <label
            class="sc-AxgMl etVWWD"
          >
            <input
              class="sc-Axmtr cIqeNF"
              data-testid="todo-completed-input"
              type="checkbox"
            />
            <div
              class="sc-AxheI bpzKbG"
            >
              <svg
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polyline
                  points="20 6 9 17 4 12"
                />
              </svg>
            </div>
          </label>
        </div>
        <div
          class="sc-AxiKw hJwCtr"
          data-testid="todo-content"
        >
          <input
            class="sc-fzozJi fOBrlk"
            data-testid="todo-content-input"
            value="find new job"
          />
        </div>
        <div
          class="sc-AxhCb fgtRIY"
        >
          <button
            class="sc-AxhUy kifWmV"
            data-testid="todo-button"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="18"
                x2="6"
                y1="6"
                y2="18"
              />
              <line
                x1="6"
                x2="18"
                y1="6"
                y2="18"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-AxjAm fwkaia"
      data-testid="todo"
    >
      <div
        class="sc-AxirZ gsjyTv"
        data-testid="todo-completed"
      >
        <label
          class="sc-AxgMl etVWWD"
        >
          <input
            class="sc-Axmtr cIqeNF"
            data-testid="todo-completed-input"
            type="checkbox"
          />
          <div
            class="sc-AxheI bpzKbG"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="20 6 9 17 4 12"
              />
            </svg>
          </div>
        </label>
      </div>
      <div
        class="sc-AxiKw hJwCtr"
        data-testid="todo-content"
      >
        <input
          class="sc-fzozJi fOBrlk"
          data-testid="todo-content-input"
          value="find new job"
        />
      </div>
      <div
        class="sc-AxhCb fgtRIY"
      >
        <button
          class="sc-AxhUy kifWmV"
          data-testid="todo-button"
        >
          <svg
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="18"
              x2="6"
              y1="6"
              y2="18"
            />
            <line
              x1="6"
              x2="18"
              y1="6"
              y2="18"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
